ESP32S3 PSRAM is not working properly: only work in octal+40M

### Describe the bug

Zephyr 4.4's spiram_test may encounter compilation and runtime errors on ESP32S3 depending on the configuration.
- Quad_psram +40M or 80M does not work properly.
- Octal_psram +80M results in a linking error during build.
- Octal_psram +40M works properly.

Zephyr 4.3 works without any problems.

### Regression

- [ ] This is a regression.

### Steps to reproduce

`west build -b esp32s3_devkitc/esp32s3/procpu zephyr/samples/boards/espressif/spiram_test/`
1. Only add `CONFIG_SPIRAM_MODE_OCT=y` => work properly
2. Do not modify any configuration items, uart console output:
> (quad_psram): PSRAM chip is not connected, or wrong PSRAM line mode
3. Only add `CONFIG_SPIRAM_SPEED_80M=y`, uart console output:
>(quad_psram): PSRAM chip is not connected, or wrong PSRAM line mode
4. Only add `CONFIG_SPIRAM_SPEED_80M=y` and `CONFIG_SPIRAM_MODE_OCT=y`, build output:
>[1/6] Linking C executable zephyr/zephyr_pre0.elf
FAILED: zephyr/zephyr_pre0.elf zephyr/zephyr_pre0.map /home/frank/work/zpro/zephyrproject/build/zephyr/zephyr_pre0.map 
: && ccache /home/frank/work/zpro/tools/zephyr-sdk-1.0.1/gnu/xtensa-espressif_esp32s3_zephyr-elf/bin/xtensa-espressif_esp32s3_zephyr-elf-gcc  -gdwarf-4 -Os zephyr/CMakeFiles/zephyr_pre0.dir/misc/empty_file.c.obj -o zephyr/zephyr_pre0.elf  zephyr/CMakeFiles/offsets.dir/./arch/xtensa/core/offsets/offsets.c.obj  -T  zephyr/linker_zephyr_pre0.cmd  -Wl,-Map,/home/frank/work/zpro/zephyrproject/build/zephyr/zephyr_pre0.map  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  zephyr/arch/common/libarch__common.a  zephyr/arch/arch/xtensa/core/libarch__xtensa__core.a  zephyr/lib/libc/picolibc/liblib__libc__picolibc.a  zephyr/lib/libc/common/liblib__libc__common.a  zephyr/subsys/portability/posix/c_lib_ext/libsubsys__portability__posix__c_lib_ext.a  zephyr/drivers/interrupt_controller/libdrivers__interrupt_controller.a  zephyr/drivers/clock_control/libdrivers__clock_control.a  zephyr/drivers/console/libdrivers__console.a  zephyr/drivers/gpio/libdrivers__gpio.a  zephyr/drivers/pinctrl/libdrivers__pinctrl.a  zephyr/drivers/serial/libdrivers__serial.a  zephyr/drivers/timer/libdrivers__timer.a  -Wl,--no-whole-archive  zephyr/kernel/libkernel.a  -L/home/frank/work/zpro/zephyrproject/build/zephyr  zephyr/arch/common/libisr_tables.a  -fuse-ld=bfd  -Wl,--gc-sections  -Wl,--build-id=none  -Wl,--sort-common=descending  -Wl,--sort-section=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -znoexecstack  -Wl,-X  -Wl,-N  -Wl,--orphan-handling=warn  -Wl,-no-pie  -Wl,--undefined=_sw_isr_table  -Wl,--undefined=_irq_vector_table  -specs=picolibc.specs  -DPICOLIBC_LONG_LONG_PRINTF_SCANF  -lgcc  -T/home/frank/work/zpro/zephyrproject/modules/hal/espressif/zephyr/esp32s3/src/linker/esp32s3.rom.alias.ld  -T/home/frank/work/zpro/zephyrproject/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.ld  -T/home/frank/work/zpro/zephyrproject/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.api.ld  -T/home/frank/work/zpro/zephyrproject/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.libgcc.ld  -T/home/frank/work/zpro/zephyrproject/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.newlib.ld  -T/home/frank/work/zpro/zephyrproject/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.version.ld  -T/home/frank/work/zpro/zephyrproject/modules/hal/espressif/zephyr/esp32s3/../../components/esp_rom/esp32s3/ld/esp32s3.rom.libc.ld  -T/home/frank/work/zpro/zephyrproject/modules/hal/espressif/zephyr/esp32s3/../../components/soc/esp32s3/ld/esp32s3.peripherals.ld  -Wl,--wrap=longjmp -L"/home/frank/work/zpro/tools/zephyr-sdk-1.0.1/gnu/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/14.3.0/space" -lc -lgcc && cd /home/frank/work/zpro/zephyrproject/build/zephyr && /usr/bin/cmake -E true
/home/frank/work/zpro/tools/zephyr-sdk-1.0.1/gnu/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/14.3.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: zephyr/libzephyr.a(mspi_timing_tuning.c.obj):(.literal.mspi_timing_enter_low_speed_mode+0x0): undefined reference to `mspi_timing_flash_config_clear_tuning_regs'
/home/frank/work/zpro/tools/zephyr-sdk-1.0.1/gnu/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/14.3.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: zephyr/libzephyr.a(mspi_timing_tuning.c.obj):(.literal.mspi_timing_enter_low_speed_mode+0x4): undefined reference to `mspi_timing_psram_config_clear_tuning_regs'
/home/frank/work/zpro/tools/zephyr-sdk-1.0.1/gnu/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/14.3.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: zephyr/libzephyr.a(mspi_timing_tuning.c.obj):(.literal.mspi_timing_enter_high_speed_mode+0x0): undefined reference to `mspi_timing_flash_config_set_tuning_regs'
/home/frank/work/zpro/tools/zephyr-sdk-1.0.1/gnu/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/14.3.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: zephyr/libzephyr.a(mspi_timing_tuning.c.obj):(.literal.mspi_timing_enter_high_speed_mode+0x4): undefined reference to `mspi_timing_psram_config_set_tuning_regs'
/home/frank/work/zpro/tools/zephyr-sdk-1.0.1/gnu/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/14.3.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: zephyr/libzephyr.a(mspi_timing_tuning.c.obj):(.literal.mspi_timing_psram_tuning+0x8): undefined reference to `mspi_timing_psram_init'
/home/frank/work/zpro/tools/zephyr-sdk-1.0.1/gnu/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/14.3.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: zephyr/libzephyr.a(mspi_timing_tuning.c.obj):(.literal.mspi_timing_psram_tuning+0xc): undefined reference to `mspi_timing_config_psram_set_tuning_regs'
/home/frank/work/zpro/tools/zephyr-sdk-1.0.1/gnu/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/14.3.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: zephyr/libzephyr.a(mspi_timing_tuning.c.obj):(.literal.mspi_timing_psram_tuning+0x10): undefined reference to `mspi_timing_config_psram_read_data'
/home/frank/work/zpro/tools/zephyr-sdk-1.0.1/gnu/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/14.3.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: zephyr/libzephyr.a(mspi_timing_tuning.c.obj):(.literal.mspi_timing_psram_tuning+0x14): undefined reference to `mspi_timing_psram_select_best_tuning_config'
/home/frank/work/zpro/tools/zephyr-sdk-1.0.1/gnu/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/14.3.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: zephyr/libzephyr.a(mspi_timing_tuning.c.obj):(.literal.mspi_timing_psram_tuning+0x18): undefined reference to `mspi_timing_psram_set_best_tuning_config'
/home/frank/work/zpro/tools/zephyr-sdk-1.0.1/gnu/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/14.3.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: zephyr/libzephyr.a(mspi_timing_tuning.c.obj):(.literal.mspi_timing_psram_tuning+0x1c): undefined reference to `mspi_timing_config_psram_prepare_reference_data'
/home/frank/work/zpro/tools/zephyr-sdk-1.0.1/gnu/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/14.3.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: zephyr/libzephyr.a(mspi_timing_tuning.c.obj):(.literal.mspi_timing_psram_tuning+0x20): undefined reference to `mspi_timing_config_psram_write_data'
/home/frank/work/zpro/tools/zephyr-sdk-1.0.1/gnu/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/14.3.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: zephyr/libzephyr.a(mspi_timing_tuning.c.obj):(.literal.mspi_timing_psram_tuning+0x24): undefined reference to `mspi_timing_get_psram_tuning_configs'
/home/frank/work/zpro/tools/zephyr-sdk-1.0.1/gnu/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/14.3.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: zephyr/libzephyr.a(mspi_timing_tuning.c.obj):(.literal.spi_timing_get_flash_timing_param+0x0): undefined reference to `mspi_timing_config_get_flash_clock_reg'
/home/frank/work/zpro/tools/zephyr-sdk-1.0.1/gnu/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/14.3.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: zephyr/libzephyr.a(mspi_timing_tuning.c.obj):(.literal.spi_timing_get_flash_timing_param+0x4): undefined reference to `mspi_timing_config_get_flash_extra_dummy'
/home/frank/work/zpro/tools/zephyr-sdk-1.0.1/gnu/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/14.3.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: zephyr/libzephyr.a(mspi_timing_tuning.c.obj):(.literal.spi_timing_get_flash_timing_param+0x8): undefined reference to `mspi_timing_config_get_cs_timing'
/home/frank/work/zpro/tools/zephyr-sdk-1.0.1/gnu/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/14.3.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: zephyr/libzephyr.a(mspi_timing_tuning.c.obj): in function `mspi_timing_enter_low_speed_mode':
/home/frank/work/zpro/zephyrproject/modules/hal/espressif/components/esp_hw_support/mspi/mspi_timing_tuning/mspi_timing_tuning.c:510:(.text.mspi_timing_enter_low_speed_mode+0x1c): undefined reference to `mspi_timing_flash_config_clear_tuning_regs'
/home/frank/work/zpro/tools/zephyr-sdk-1.0.1/gnu/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/14.3.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: /home/frank/work/zpro/zephyrproject/modules/hal/espressif/components/esp_hw_support/mspi/mspi_timing_tuning/mspi_timing_tuning.c:512:(.text.mspi_timing_enter_low_speed_mode+0x24): undefined reference to `mspi_timing_psram_config_clear_tuning_regs'
/home/frank/work/zpro/tools/zephyr-sdk-1.0.1/gnu/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/14.3.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: zephyr/libzephyr.a(mspi_timing_tuning.c.obj): in function `mspi_timing_enter_high_speed_mode':
/home/frank/work/zpro/zephyrproject/modules/hal/espressif/components/esp_hw_support/mspi/mspi_timing_tuning/mspi_timing_tuning.c:539:(.text.mspi_timing_enter_high_speed_mode+0x1c): undefined reference to `mspi_timing_flash_config_set_tuning_regs'
/home/frank/work/zpro/tools/zephyr-sdk-1.0.1/gnu/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/14.3.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: /home/frank/work/zpro/zephyrproject/modules/hal/espressif/components/esp_hw_support/mspi/mspi_timing_tuning/mspi_timing_tuning.c:544:(.text.mspi_timing_enter_high_speed_mode+0x24): undefined reference to `mspi_timing_psram_config_set_tuning_regs'
/home/frank/work/zpro/tools/zephyr-sdk-1.0.1/gnu/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/14.3.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: zephyr/libzephyr.a(mspi_timing_tuning.c.obj): in function `mspi_timing_psram_tuning':
/home/frank/work/zpro/zephyrproject/modules/hal/espressif/components/esp_hw_support/mspi/mspi_timing_tuning/mspi_timing_tuning.c:411:(.text.mspi_timing_psram_tuning+0x1d): undefined reference to `mspi_timing_config_psram_prepare_reference_data'
/home/frank/work/zpro/tools/zephyr-sdk-1.0.1/gnu/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/14.3.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: /home/frank/work/zpro/zephyrproject/modules/hal/espressif/components/esp_hw_support/mspi/mspi_timing_tuning/mspi_timing_tuning.c:412:(.text.mspi_timing_psram_tuning+0x29): undefined reference to `mspi_timing_config_psram_write_data'
/home/frank/work/zpro/tools/zephyr-sdk-1.0.1/gnu/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/14.3.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: /home/frank/work/zpro/zephyrproject/modules/hal/espressif/components/esp_hw_support/mspi/mspi_timing_tuning/mspi_timing_tuning.c:427:(.text.mspi_timing_psram_tuning+0x80): undefined reference to `mspi_timing_get_psram_tuning_configs'
/home/frank/work/zpro/tools/zephyr-sdk-1.0.1/gnu/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/14.3.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: zephyr/libzephyr.a(mspi_timing_tuning.c.obj): in function `spi_timing_get_flash_timing_param':
/home/frank/work/zpro/zephyrproject/modules/hal/espressif/components/esp_hw_support/mspi/mspi_timing_tuning/mspi_timing_tuning.c:592:(.text.spi_timing_get_flash_timing_param+0x3): undefined reference to `mspi_timing_config_get_flash_clock_reg'
/home/frank/work/zpro/tools/zephyr-sdk-1.0.1/gnu/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/14.3.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: /home/frank/work/zpro/zephyrproject/modules/hal/espressif/components/esp_hw_support/mspi/mspi_timing_tuning/mspi_timing_tuning.c:638:(.text.spi_timing_get_flash_timing_param+0xe): undefined reference to `mspi_timing_config_get_flash_extra_dummy'
/home/frank/work/zpro/tools/zephyr-sdk-1.0.1/gnu/xtensa-espressif_esp32s3_zephyr-elf/bin/../lib/gcc/xtensa-espressif_esp32s3_zephyr-elf/14.3.0/../../../../xtensa-espressif_esp32s3_zephyr-elf/bin/ld.bfd: /home/frank/work/zpro/zephyrproject/modules/hal/espressif/components/esp_hw_support/mspi/mspi_timing_tuning/mspi_timing_tuning.c:645:(.text.spi_timing_get_flash_timing_param+0x1a): undefined reference to `mspi_timing_config_get_cs_timing'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.


### Relevant log output

```shell

```

### Impact

Annoyance – Minor irritation; no significant impact on usability or functionality.

### Environment

_No response_

### Additional Context

_No response_

--- a/samples/boards/espressif/spiram_test/sample.yaml
+++ b/samples/boards/espressif/spiram_test/sample.yaml
@@ -1,13 +1,23 @@
 sample:
   description: Application to test memory allocation from SPIRAM
   name: spiram_test
+common:
+  harness: console
+  harness_config:
+    type: multi_line
+    regex:
+      - "SPIRAM mem test pass"
+      - "Internal mem test pass"
 tests:
   sample.board.esp32.spiram:
     platform_allow: esp32_devkitc/esp32/procpu
     tags: esp32
-    harness: console
-    harness_config:
-      type: multi_line
-      regex:
-        - "SPIRAM mem test pass"
-        - "Internal mem test pass"
+  sample.board.esp32s3.spiram:
+    platform_allow: esp32s3_devkitc/esp32s3/procpu
+    tags: esp32s3
+  sample.board.esp32s3.spiram.octal_80m:
+    platform_allow: esp32s3_devkitc/esp32s3/procpu
+    tags: esp32s3
+    extra_configs:
+      - CONFIG_SPIRAM_SPEED_80M=y
+      - CONFIG_SPIRAM_MODE_OCT=y

--- a/soc/espressif/esp32s3/default.ld
+++ b/soc/espressif/esp32s3/default.ld
@@ -464,6 +464,7 @@ SECTIONS
     *libzephyr.a:systimer.*(.literal .literal.* .text .text.*)
     *libzephyr.a:mspi_timing_config.*(.literal .literal.* .text .text.*)
     *libzephyr.a:mspi_timing_tuning.*(.literal .literal.* .text .text.*)
+    *libzephyr.a:mspi_timing_by_mspi_delay.*(.literal .literal.* .text .text.*)
     *libzephyr.a:regi2c_ctrl.*(.literal .text .literal.* .text.*)
     *(.literal.sar_periph_ctrl_power_enable .text.sar_periph_ctrl_power_enable)
 
@@ -709,6 +710,7 @@ SECTIONS
     *libzephyr.a:systimer.*(.rodata .rodata.*)
     *libzephyr.a:mspi_timing_config.*(.rodata .rodata.*)
     *libzephyr.a:mspi_timing_tuning.*(.rodata .rodata.*)
+    *libzephyr.a:mspi_timing_by_mspi_delay.*(.rodata .rodata.*)
     *(.rodata.sar_periph_ctrl_power_enable)
 
     /* [mapping:soc_pm] */

--- a/west.yml
+++ b/west.yml
@@ -172,7 +172,7 @@ manifest:
       groups:
         - hal
     - name: hal_espressif
-      revision: b7953b8019361d09e613f7011d2ccc41b984d087
+      revision: 15789ec576bdb3397c6fc39542b0d9b534d98d2d
       path: modules/hal/espressif
       west-commands: west/west-commands.yml
       groups:

